src/main/java/uk/ac/ebi/ampt2d/accession/common/accessioning/BasicAccessioningService.java

```diff
@@ -0,0 +1,136 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package uk.ac.ebi.ampt2d.accession.common.accessioning;
+
+import org.springframework.beans.factory.InitializingBean;
+import uk.ac.ebi.ampt2d.accession.common.generators.AccessionGenerator;
+import uk.ac.ebi.ampt2d.accession.common.persistence.DatabaseService;
+import uk.ac.ebi.ampt2d.accession.common.utils.DigestFunction;
+import uk.ac.ebi.ampt2d.accession.common.utils.HashingFunction;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * A service that provides accessions for objects
+ *
+ * @param <MODEL>
+ * @param <ACCESSION>
+ * @param <HASH>
+ */
+public class BasicAccessioningService<MODEL, HASH, ACCESSION> implements InitializingBean,
+        AccessioningService<MODEL, ACCESSION> {
+
+    private AccessionGenerator<MODEL, ACCESSION> accessionGenerator;
+
+    private DatabaseService<MODEL, HASH, ACCESSION> dbService;
+
+    private final DigestFunction<MODEL> digestFunction;
+
+    private HashingFunction<HASH> hashingFunction;
+
+    /**
+     * Basic accessioning service constructor, requires an accession generator, a database service and a hashing
+     * function, it also allows the registering of a init function.
+     *
+     * @param accessionGenerator
+     * @param dbService
+     * @param hashingFunction
+     */
+    public BasicAccessioningService(AccessionGenerator<MODEL, ACCESSION> accessionGenerator,
+                                    DatabaseService<MODEL, HASH, ACCESSION> dbService,
+                                    DigestFunction<MODEL> digestFunction,
+                                    HashingFunction<HASH> hashingFunction) {
+        this.accessionGenerator = accessionGenerator;
+        this.dbService = dbService;
+        this.digestFunction = digestFunction;
+        this.hashingFunction = hashingFunction;
+    }
+
+    /**
+     * Get accessions for a list of messages. It looks for the object's accessions in a repository, and it they don't
+     * exist, generate new ones, storing them in the repository
+     *
+     * @param messages
+     * @return
+     */
+    @Override
+    public Map<ACCESSION, MODEL> getOrCreateAccessions(List<? extends MODEL> messages) {
+        Map<HASH, MODEL> hashToMessages = mapHashOfMessages(messages);
+        Map<HASH, ACCESSION> existingAccessions = dbService.getExistingAccessions(hashToMessages.keySet());
+        Map<HASH, MODEL> newMessages = filterNotExistingAccessions(hashToMessages, existingAccessions);
+
```

Comment:
Map<ACCESSION, MODEL> existingAccessions = getAccessions(messages);
 Set<MODEL> existingModels = new HashSet<>(existingAccessions.values());
 Map<HASH, MODEL> newMessages = messages.stream().filter(msg -> !existingModels.contains(msg))
                .collect(Collectors.toMap(digestFunction.andThen(hashingFunction), msg -> msg, (o, n) -> o));

How about this , this would reduce other private functions and make use of the library. Don't mind if it doesn't make any sense , as the code is good right now as well.

Reply:
But for that I would need to make mandatory that people makes a comparison function in their entities and dtos that works even when they are comparing between themselves. Small private code vs that per each class... I stand on the current design.